pkg/envoy/xds/watcher.go

```diff
@@ -118,6 +118,17 @@ func (w *ResourceWatcher) WatchResources(ctx context.Context, typeURL string, la
 
 	for ctx.Err() == nil && res == nil {
 		w.versionLocker.Lock()
+		// If the client ACKed a version that we have never sent back, this
+		// indicates that this server restarted but the client survived and had
+		// received a higher version number from the previous server instance.
+		// Bump the resource set's version number to match the client's and
+		// send a response immediately.
+		if waitForVersion && w.version < waitVersion {
+			w.versionLocker.Unlock()
+			w.resourceSet.EnsureVersion(typeURL, waitVersion+1)
```

Comment:
I would have expected this logic to be closer to the code that determines that an ACK was received. Is this placed here for testing reasons?

Reply:
The version of resources is encapsulated behind the ResourceSet interface. Server / Stream don't have direct access to resource sets. They only access them through watchers, which encapsulate the resource sets. The design in this PR preserves that encapsulation and separation of concerns.

And it is the watcher which decides when to reply to a request.
So it seems like the best place to put the logic about version updates and whether to reply immediately to a request when it has a more recent version.